File: 프로그래머스/2/42577. 전화번호 목록/전화번호 목록.py
class Node():
    def __init__(self):
        self.isEnd = False
        self.children = {}

class Trie():
    def __init__(self):
        self.root = Node()
    
    def insert(self, string):
        root = self.root
        
        for char in string:
            if char not in root.children:
                root.children[char] = Node()
            
            root = root.children[char]
            
            if root.isEnd:
                return True
            
        root.isEnd = True
        
        return bool(root.children)
    

def solution(phone_book):
    trie = Trie()
    for item in phone_book:
        if trie.insert(item):
            return False
        
    return True

File: 프로그래머스/2/42577. 전화번호 목록/전화번호 목록.js
class Node {
    isEnd = false
    children = Array(10)
}

class Trie {
    constructor() {
        this.pool = [new Node()]
    }

    insert(string) {
        let cur = this.pool[0]

        for (let i = 0; i < string.length; i++) {
            const char = string[i]
            
            if (!cur.children[Number(char)]) {
                cur.children[Number(char)] = this.pool.length
                this.pool.push(new Node())
            } else if (i === string.length - 1) {
                return true
            }

            cur = this.pool[cur.children[Number(char)]]

            if (cur.isEnd){
                return true
            }
        }

        cur.isEnd = true

        return false
    }


}

function solution(phone_book) {
    const trie = new Trie()

    for (const item of phone_book) {
        if (trie.insert(item)) {
            return false
        }
    }

    return true
}